### ISSUE TITLE
integrations: cockroachdb

### ISSUE THREAD

**Issue by nickbp** [opstrace/opstrace#1278]
Investigate and build a new opstrace integration for monitoring a cockroachdb cluster.

This could be done by manually deploying a prometheus scraper as in [our recent blog post](https://opstrace.com/blog/collecting-metrics-from-cockroachdb) where we do this manually, but could we get away with instead using [datadog-agent](https://docs.datadoghq.com/agent/docker/) and pointing it to [opstrace ddapi](https://opstrace.com/blog/introducing-datadog-compatible-http-api)?

It'd make sense to pick a method that would also be straightforward to integrate with e.g. postgresql/mysql so that we can more easily add integrations for those and other similar packages later on. It looks like datadog-agent has built-in support for [postgres](https://docs.datadoghq.com/integrations/postgres/?tab=kubernetes) and [mysql](https://docs.datadoghq.com/integrations/mysql/?tab=kubernetes) with some requirements around creating a DB user for the agent to query stats against.

See also the [cockroach repo `/monitoring` directory](https://github.com/cockroachdb/cockroach/tree/master/monitoring) containing sample dashboards and prometheus scrape config.

**Comment by triclambert** [comment]
See also:  https://github.com/opstrace/opstrace/issues/429

**Comment by nickbp** [comment]
From a quick discussion earlier it looks like [grafana-agent](https://grafana.com/docs/grafana-cloud/agent/) would be better just because it would allow sending the data straight to cortex rather than via the ddapi conversion. And it has docs for integrating with [cockroachdb](https://grafana.com/docs/grafana-cloud/integrations/integrations/integration-cockroachdb/) as well as the examples of postgres and mysql.

Their integrations docs are written with Grafana Cloud in mind which can make it a little confusing. However the agent is where the support is implemented:
> When the Grafana Cloud team builds Integration for users to install, code is also added to the agent, so that the agent and Grafana Cloud send and receive metrics for the data source. You can’t see which Integrations are supported for your agent version. This makes it important to update your agent version any time you want to use a new Integration you have not previously enabled.

---

Separately, unlike with the existing K8s integrations, users could be running cockroachdb from within K8s, on the host system directly, or within docker on a host. So we'd want to provide two or three different set of instructions based on where it's running.

One way of presenting this would be to just have different tabs/sections in the install guide, where the user picks the method after creating the integration. This would work well if the integration config options to display at setup are largely uniform across the install methods.

Another way of presenting it would be to prompt the user to select the install method up-front when the integration is being created. This would work well in cases where the integration config itself will vary based on the install type (for example: only show a K8s namespace text box if the user selects K8s as the install type).

But for now we can just start with one install type for the first pass and then add more options in later revisions.

**Comment by nickbp** [comment]
Also, unlike with the existing K8s integrations, users could be running cockroachdb from within K8s, on the host system directly, or within docker on a host. So we'd want to provide two or three different set of instructions based on where it's running.

One way of presenting this would be to just have different tabs/sections in the install guide, where the user picks the method after creating the integration. This would work well if the integration config options to display at setup are largely uniform across the install methods.

Another way of presenting it would be to prompt the user to select the install method up-front when the integration is being created. This would work well in cases where the integration config itself will vary based on the install type (for example: only show a K8s namespace text box if the user selects K8s as the install type).

But for now we can just start with one install type for the first pass and then add more options in later revisions.

**Comment by nickbp** [comment]
So at this point I've got a running cockroachdb instance on my workstation (3 nodes on separate ports with certs configured) and I've been testing out various ways of scraping the instances using grafana-agent. The grafana-agent itself has been working well and I think it would be a good replacement for the scraper Prometheuses that we're using today both for benchmarks as well as for the k8sMetrics integration, since it's ultimately a stripped down version of prometheus that just does scraping+`remote_write` and that's what we're using the scrapers for. One downside is that it isn't as debuggable, only having a `/metrics` endpoint with scrape stats when a full prometheus has the nice dashboard.

I've thus far been testing just specifying the nodes via `static_config` and providing the generated CA certificate and that's worked fine. So for a non-K8s cockroach deployment I think we can just provide an example of that.

There are a few decisions that I've wrestling with after some testing, and they are a bit inter-related:
- How to ensure that the `instance` label on the different nodes is correct. With a single scraper for all nodes it should be possible to just use the target hostname, whereas for multiple scrapers (each at `localhost:8080` respectively) that wouldn't really work.
- Whether to go with a single scraper for multiple DB nodes, or to have multiple scrapers sidecar-style. I'm leaning towards single scraper because:
    - It's a bit less work to get up and running (especially for non-K8s)
    - Avoids the above `instance` target labeling issue since the targets would be different
    - In the K8s case, changes don't involve needing to do a rolling cockroach pod restart to update sidecars

So in summary here's the plan:
- Use grafana-agent as mentioned before
- K8s: single-pod Deployment of the agent configured with the tenant auth token that uses service discovery to find the cockroachdb pods. Functionally equivalent to the k8sMetrics integration except with the user specifying the name (and port?) of the service to look up.
- Standalone: single-process scraper pointed to the nodes as retrieved via a manual `cockroach node status` command (specifically something like `cockroach node status --format tsv --certs-dir ./certs/ | awk '{print $2}' | tail -n +2`). We'd link to the grafana-agent binary to download.

**Comment by MatApple** [comment]
Sounds great @nickbp - I like where you landed with a single scraper grafana-agent + user-specified service name + port. Think that makes a lot of sense

**Comment by nickbp** [comment]
(This was done via #1321 )
